source/Fido.Functional.pas

```diff
@@ -0,0 +1,341 @@
+(*
+ * Copyright 2022 Mirko Bianco (email: writetomirko@gmail.com)
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *)
+
+unit Fido.Functional;
+
+interface
+
+uses
+  System.SysUtils,
+  System.Threading,
+  System.Rtti,
+
+  Spring,
+
+  Fido.Exceptions,
+  Fido.Boxes;
+
+type
+  // https://adit.io/posts/2013-04-17-functors,_applicatives,_and_monads_in_pictures.html
+  //Functor: applies a function to a wrapped value
+  //Applicative: applies wrapped function to a wrapped value
+  //Monad: applies a function(which returns a wrapped value) to a wrapped value
+
+  EFunctionalContext = class(EFidoException);
+
+  OnFailureEvent<T> = reference to function(const Value: TObject): T;
+  OnSuccessEvent<T> = reference to function(const Value: T): T;
+
+  Context<T> = record
+  public type
+    FunctorFunc<TOut> = reference to function(const Value: T): TOut;
+    MonadFunc<TOut> = reference to function(const Value: T): Context<TOut>;
+    FunctorProc = reference to procedure(const Value: T);
+  private
+    FAssigned: string;
```

Comment:
Why is FAssigned a string, while it is used as a boolean?

Reply:
ehehehe, my young friend....

Because being a string (a managed type) it will be automatically set as '' upon initialization.
Context<T> being a record, it doesn't like parameterless constructors.
This is a trick to setup a flag using a string, without a constructor.

Reply:
Ah that's right! I forgot bools are not a managed type.